src/mapgen_valleys.cpp

```diff
@@ -897,101 +783,150 @@ void MapgenValleys::dustTopNodes()
 
 	v3s16 em = vm->m_area.getExtent();
 
-	s16 base_water_chance = 0;
-	if (water_features < 11)
-		base_water_chance = ceil(MAX_MAP_GENERATION_LIMIT / (water_features * 1000));
-
-	if (max_stone_y >= node_min.Y) {
-		u32 index_2d = 0;
-		u32 index_3d = 0;
-		for (s16 z = node_min.Z; z <= node_max.Z; z++)
-		for (s16 x = node_min.X; x <= node_max.X; x++, index_2d++) {
-			bool air_above = false;
-			//bool underground = false;
-			u32 index_data = vm->m_area.index(x, node_max.Y + 1, z);
-
-			index_3d = (z - node_min.Z) * zstride + (csize.Y + 1) * ystride + (x - node_min.X);
-
-			// Dig caves on down loop to check for air above.
-			for (s16 y = node_max.Y + 1;
-					y >= node_min.Y - 1;
-					y--, index_3d -= ystride, vm->m_area.add_y(em, index_data, -1)) {
-				float terrain = noise_terrain_height->result[index_2d];
-
-				// Saves some time and prevents removing above ground nodes.
-				if (y > terrain + 1) {
-					air_above = true;
-					continue;
-				}
+	// Cave blend distance near YMIN, YMAX
+	const float massive_cave_blend = 128.f;
+	// noise threshold for massive caves
+	const float massive_cave_threshold = 0.6f;
+	// mct: 1 = small rare caves, 0.5 1/3rd ground volume, 0 = 1/2 ground volume.
 
-				content_t c = vm->m_data[index_data].getContent();
-				bool n1 = (fabs(noise_simple_caves_1->result[index_3d]) < 0.07f);
-				bool n2 = (fabs(noise_simple_caves_2->result[index_3d]) < 0.07f);
+	float yblmin = -map_gen_limit + massive_cave_blend * 1.5f;
+	float yblmax = massive_cave_depth - massive_cave_blend * 1.5f;
+	bool made_a_big_one = false;
+	float tcave_cache[csize.Y + 2];
 
-				// River water is (foolishly) not set as ground content
-				// in the default game. This can produce strange results
-				// when a cave undercuts a river. However, that's not for
-				// the mapgen to correct. Fix it in lua.
+	// Cache the tcave values as they only vary by altitude.
+	if (node_max.Y <= massive_cave_depth) {
+		for (s16 y = node_min.Y - 1; y <= node_max.Y + 1; y++) {
+			float tcave = massive_cave_threshold;
 
-				if (c == CONTENT_AIR) {
-					air_above = true;
-				} else if (n1 && n2 && ndef->get(c).is_ground_content) {
-					// When both n's are true, we're in a cave.
-					vm->m_data[index_data] = n_air;
-					air_above = true;
-				} else if (air_above
-						&& (c == c_stone || c == c_sandstone || c == c_desert_stone)) {
-					// At the cave floor
-					s16 sr = ps.next() & 1023;
-					u32 j = index_data;
-					vm->m_area.add_y(em, j, 1);
-
-					if (sr > (terrain - y) * 25) {
-						// Put dirt in caves near the surface.
-						Biome *biome = (Biome *)bmgr->getRaw(biomemap[index_2d]);
+			if (y < yblmin)
+				tcave += pow(((yblmin - y) / massive_cave_blend), 2);
+			else if (y > yblmax)
+				tcave += pow(((y - yblmax) / massive_cave_blend), 2);
+
+			tcave_cache[y - node_min.Y + 1] = tcave;
+		}
+	}
+
+	// This allows random lava spawns to be less common at the surface.
+	s16 lava_chance = pow(lava_features_lim, 3)
```

Comment:
No problem. However, "We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil" -- Knuth